src/env.rs

```diff
@@ -0,0 +1,278 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An environment which is passed downward into the widget tree.
+
+// TODO: remove this
+#![allow(unused)]
+
+use std::collections::HashMap;
+use std::ops::{Deref, Index};
+
+use crate::piet::Color;
+
+use crate::Data;
+
+// Discussion question: since druid is pretty rigorously single-threaded,
+// maybe we should use Rc instead of Arc.
+use std::sync::Arc;
+
+/// An environment passed down through all widget traversals.
+///
+/// All widget methods have access to an environment, and it is passed
+/// downwards during traversals.
+///
+/// A widget can retrieve theme parameters (colors, dimensions, etc.). In
+/// addition, it can pass custom data down to all descendants. An important
+/// example of the latter is setting a value for enabled/disabled status
+/// so that an entire subtree can be disabled ("grayed out") with one
+/// setting.
+#[derive(Clone)]
+pub struct Env(Arc<EnvImpl>);
+
+struct EnvImpl {
+    theme: EnvValue,
+    // This is always a map, maybe should be a map here.
+    values: EnvValue,
+}
+
+/// A value that can be stored in an environment.
+///
+/// This is an enum with a variety of common types.
+#[derive(Clone)]
+pub enum EnvValue {
+    Float(f64),
+    Bool(bool),
+    Color(Color),
+    String(String),
+    List(Arc<Vec<EnvValue>>),
+    Map(Arc<HashMap<String, EnvValue>>),
+    Null,
```

Comment:
This sets of my spidey sense. What does this give us that we can't do with `Option`? Having indexing work is a very questionable gain if the thing we get back may be invalid; if something can be missing, forcing the user to handle an `Option` feels better, because that's just a common Rust pattern.